irohad/torii/impl/command_service.cpp

```diff
@@ -14,72 +14,71 @@ See the License for the specific language governing permissions and
-          break;
-        case iroha::model::TransactionResponse::COMMITTED:
-          res->second.set_tx_status(
-              iroha::protocol::TxStatus::COMMITTED);
-          break;
-        case iroha::model::TransactionResponse::ON_PROCESS:
-          res->second.set_tx_status(
-              iroha::protocol::TxStatus::ON_PROCESS);
-          break;
-        case iroha::model::TransactionResponse::NOT_RECEIVED:
-          res->second.set_tx_status(
-              iroha::protocol::TxStatus::NOT_RECEIVED);
-          break;
-      }
-
-      this->handler_map_.insert({iroha_response->tx_hash, res->second});
-    });
+    tx_processor_->transactionNotifier().subscribe(
+        [this](std::shared_ptr<TransactionResponse> iroha_response) {
+          // Find response in handler map
+          auto res = this->handler_map_.find(iroha_response->tx_hash);
+          if (res == this->handler_map_.end()) {
+            ToriiResponse response;
+            response.set_tx_hash(iroha_response->tx_hash);
+            response.set_tx_status(TxStatus::NOT_RECEIVED);
+            this->handler_map_.insert({iroha_response->tx_hash, response});
+            return;
+          }
+          switch (iroha_response->current_status) {
+            case TransactionResponse::STATELESS_VALIDATION_FAILED:
+              res->second.set_tx_status(TxStatus::STATELESS_VALIDATION_FAILED);
+              break;
+            case TransactionResponse::STATELESS_VALIDATION_SUCCESS:
+              res->second.set_tx_status(TxStatus::STATELESS_VALIDATION_SUCCESS);
+              break;
+            case TransactionResponse::STATEFUL_VALIDATION_FAILED:
+              res->second.set_tx_status(TxStatus::STATEFUL_VALIDATION_FAILED);
+              break;
+            case TransactionResponse::STATEFUL_VALIDATION_SUCCESS:
+              res->second.set_tx_status(TxStatus::STATEFUL_VALIDATION_SUCCESS);
+              break;
+            case TransactionResponse::COMMITTED:
+              res->second.set_tx_status(TxStatus::COMMITTED);
+              break;
+            case TransactionResponse::ON_PROCESS:
+              res->second.set_tx_status(TxStatus::ON_PROCESS);
+              break;
+            case TransactionResponse::NOT_RECEIVED:
+            default:
+              res->second.set_tx_status(TxStatus::NOT_RECEIVED);
+              break;
+          }
+
+          this->handler_map_.insert({iroha_response->tx_hash, res->second});
+        });
   }
 
-  void CommandService::ToriiAsync(iroha::protocol::Transaction const &request,
+  void CommandService::ToriiAsync(Transaction const &request,
                                   google::protobuf::Empty &empty) {
     auto iroha_tx = pb_factory_->deserialize(request);
 
     auto tx_hash = iroha::hash(*iroha_tx).to_string();
 
-    iroha::protocol::ToriiResponse response;
-    response.set_tx_status(iroha::protocol::TxStatus::ON_PROCESS);
+    ToriiResponse response;
+    response.set_tx_hash(tx_hash);
```

Comment:
Insertion of hash and status looks like a separated function, maybe refactor it?